From 48715d2bd54ea400d16107c1f95a483be5c94e2c Mon Sep 17 00:00:00 2001
From: Panu Matilainen <pmatilai@redhat.com>
Date: Wed, 19 Nov 2014 15:52:49 -0500
Subject: [net] ipv4: Fix incorrect error code when adding an unreachable route

Message-id: <0cd4fce9dffbb45087167e59c5c9678b1cc107ca.1416412130.git.pmatilai@redhat.com>
Patchwork-id: 100323
O-Subject: [RHEL7 PATCH net] ipv4: Fix incorrect error code when adding an unreachable route
Bugzilla: 1165552
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Jesper Brouer <jbrouer@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1165552
Upstream: 49dd18ba4615eaa72f15c9087dea1c2ab4744cf5 (davem/net.git)
Tested: Locally in VM
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=8267562

commit 49dd18ba4615eaa72f15c9087dea1c2ab4744cf5
Author: Panu Matilainen <pmatilai@redhat.com>
Date:   Fri Nov 14 13:14:32 2014 +0200

    ipv4: Fix incorrect error code when adding an unreachable route

    Trying to add an unreachable route incorrectly returns -ESRCH if
    if custom FIB rules are present:

    [root@localhost ~]# ip route add 74.125.31.199 dev eth0 via 1.2.3.4
    RTNETLINK answers: Network is unreachable
    [root@localhost ~]# ip rule add to 55.66.77.88 table 200
    [root@localhost ~]# ip route add 74.125.31.199 dev eth0 via 1.2.3.4
    RTNETLINK answers: No such process
    [root@localhost ~]#

    Commit 83886b6b636173b206f475929e58fac75c6f2446 ("[NET]: Change "not found"
    return value for rule lookup") changed fib_rules_lookup()
    to use -ESRCH as a "not found" code internally, but for user space it
    should be translated into -ENETUNREACH. Handle the translation centrally in
    ipv4-specific fib_lookup(), leaving the DECnet case alone.

    On a related note, commit b7a71b51ee37d919e4098cd961d59a883fd272d8
    ("ipv4: removed redundant conditional") removed a similar translation from
    ip_route_input_slow() prematurely AIUI.

    Fixes: b7a71b51ee37 ("ipv4: removed redundant conditional")
    Signed-off-by: Panu Matilainen <pmatilai@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Panu Matilainen <pmatilai@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/fib_rules.c b/net/ipv4/fib_rules.c
index 26aa65d..c5c8b24 100644
--- a/net/ipv4/fib_rules.c
+++ b/net/ipv4/fib_rules.c
@@ -62,6 +62,10 @@ int __fib_lookup(struct net *net, struct flowi4 *flp, struct fib_result *res)
  else
   res->tclassid = 0;
 #endif
+
+ if (err == -ESRCH)
+  err = -ENETUNREACH;
+
  return err;
 }
 EXPORT_SYMBOL_GPL(__fib_lookup);
-- 
1.7.1